code/datums/components/combo_attacks.dm

```diff
@@ -0,0 +1,103 @@
+	var/list/combo_list = list()
+	/// A list of strings containing the ways to do combos, for examines.
+	var/list/combo_strings = list()
+
+/datum/component/combo_attacks/Initialize(combos, examine_message, reset_message, max_combo_length, work_on_corpses = FALSE, work_on_self = FALSE, leniency_time = 5 SECONDS)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+	combo_list = combos
+	for(var/combo in combo_list)
+		var/list/combo_specifics = combo_list[combo]
+		var/step_string = english_list(combo_specifics[COMBO_STEPS])
+		combo_strings += span_notice("<b>[combo]</b> - [step_string]")
+	src.examine_message = examine_message
+	src.reset_message = reset_message
+	src.max_combo_length = max_combo_length
+	src.work_on_corpses = work_on_corpses
+	src.work_on_self = work_on_self
+	src.leniency_time = leniency_time
+
+/datum/component/combo_attacks/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE_MORE, PROC_REF(on_examine_more))
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK_SELF, PROC_REF(on_attack_self))
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, PROC_REF(on_drop))
+	RegisterSignal(parent, COMSIG_ITEM_ATTACK, PROC_REF(on_attack))
+
+/datum/component/combo_attacks/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_PARENT_EXAMINE, COMSIG_PARENT_EXAMINE_MORE, COMSIG_ITEM_ATTACK_SELF, COMSIG_ITEM_DROPPED, COMSIG_ITEM_ATTACK))
+
+/datum/component/combo_attacks/proc/on_examine(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	if(!examine_message)
+		return
+	examine_list += examine_message
+
+/datum/component/combo_attacks/proc/on_examine_more(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	examine_list += combo_strings
+
+/datum/component/combo_attacks/proc/on_attack_self(obj/item/source, mob/user)
+	SIGNAL_HANDLER
+
+	reset_inputs(user, deltimer = TRUE)
+
+/datum/component/combo_attacks/proc/on_drop(datum/source, mob/dropper)
+	SIGNAL_HANDLER
+
+	reset_inputs(user = null, deltimer = TRUE)
+
+/datum/component/combo_attacks/proc/check_input(mob/living/target, mob/user)
+	for(var/combo in combo_list)
+		var/list/combo_specifics = combo_list[combo]
+		if(compare_list(input_list, combo_specifics[COMBO_STEPS]))
+			INVOKE_ASYNC(parent, combo_specifics[COMBO_PROC], target, user)
+			return TRUE
+	return FALSE
+
+/datum/component/combo_attacks/proc/reset_inputs(mob/user, deltimer)
+	var/atom/atom_parent = parent
+	input_list.Cut()
+	if(user)
+		atom_parent.balloon_alert(user, reset_message)
+	if(deltimer && timerid)
+		deltimer(timerid)
+
+/datum/component/combo_attacks/proc/on_attack(datum/source, mob/living/target, mob/user, click_parameters)
+	SIGNAL_HANDLER
+
+	if((target.stat == DEAD  && !work_on_corpses) || (target == user && !work_on_self) || HAS_TRAIT(user, TRAIT_PACIFISM))
```

Comment:
This could be made even more generic by making all these checks into a proc ref.